### ISSUE TITLE
emacsGcc deffered compliation fails

### ISSUE THREAD

**Issue by akirakyle** [nix-community/emacs-overlay#33]
First of all thanks for packaging emacsGcc!

Now that libgccjit is in unstable I was able to compile emacsGcc just find and so far it's been working well however emacs crashed when I tried setting `(setq comp-deferred-compilation t)` first thing in my .emacs. It looks like the immediate cause is due to all the resulting `*.eln` files being empty. When I try compiling any file without loading it using `native-compile-async` I see the following error for every file compiled:
```
aarch64-unknown-linux-gnu-gcc-9.3.0: fatal error: cannot execute ‘as’: execvp: No such file or directory compilation terminated.
```
I tried adding binutils to my environment but than that leads to errors like:
```
/nix/store/bmjqkgnj34c0k1bsvralz0dr6xvyfgzb-binutils-2.31.1/bin/ld: cannot find crti.o: No such file or directory
/nix/store/bmjqkgnj34c0k1bsvralz0dr6xvyfgzb-binutils-2.31.1/bin/ld: cannot find crtbeginS.o: No such file or directory
/nix/store/bmjqkgnj34c0k1bsvralz0dr6xvyfgzb-binutils-2.31.1/bin/ld: cannot find -lgcc
Native elisp load failed: "/home/akyle/.emacs.d/straight/build/magit/eln-aarch64-unknown-linux-gnu-63761565f6c9e3bd/magit.eln", "/home/akyle/.emacs.d/straight/build/magit/eln-aarch64-unknown-linux-gnu-63761565f6c9e3bd/magit.eln: file too short"
Fatal error 11: Segmentation fault
Backtrace:
[ommitted as it's just memory addresses]
```
I tried adding libgccjit to my environment but that gave me the error:
```
error: getting status of '/nix/store/kq3hwnp0bdia9ld9445fy829laf46vaf-libgccjit-9.3.0/lib/lib': Too many levels of symbolic links
```

**Comment by mjlbach** [comment]
This is because gcc can't find the libgccjit libraries at compilation time. If you want that, you need to update LIBRARY_PATH to point at the right libraries. If you evaluate the following in nix-repl, you should get the right paths (then you can update with LIBRARY_PATH=...)

```
pkgs.lib.getLib pkgs.stdenv.cc.cc + /lib
pkgs.lib.getLib pkgs.stdenv.glibc + /lib 
pkgs.lib.getLib pkgs.libgccjit + /lib/gcc/x86_64-unknown-linux-gnu/9.3.0
```

**Comment by adisbladis** [comment]
To add to what @mjlbach is saying, I don't want to add `LIBRARY_PATH` in a wrapper as it pollutes the development environment, instead I'd like to set it for the native-comp subprocess.

**Comment by akirakyle** [comment]
@mjlbach Thanks! That appears to have done the trick! This should probably be documented somewhere.

@adisbladis That makes sense. I wonder if the overlay should include some startup lisp which adds a hook to the native comp function to set the `LIBRARY_PATH` and ensure binutils is present just for the native-comp subprocess so that emacs' environment isn't polluted but a user can also safely keep `comp-deferred-compilation` set?

**Comment by adisbladis** [comment]
@akirakyle Yes, that's what I've been thinking too.

**Comment by AndreaCorallo** [comment]
I think you guys could use `comp-async-env-modifier-form` to execute the lisp you want in the environment of the async compilation process.  Is not much tested but sounds like what you need.

Note the synchronous compilation will still be crashy in case is used.

**Comment by mjlbach** [comment]
@AndreaCorallo Thanks, this seems to work for me:
```
(setq comp-async-env-modifier-form '((setenv "LIBRARY_PATH"
                                             (concat
        "/nix/store/wfvfxj1w9gj21nbai6rsyhwvkbqf59w0-libgccjit-9.3.0/lib/gcc/x86_64-unknown-linux-gnu/9.3.0:"
        "/nix/store/a6z7ighixg7gb6krf9k60ylgmahij63x-gcc-9.3.0-lib/lib:"
        "/nix/store/jx19wa4xlh9n4324xdl9rjnykd19mmq3-glibc-2.30/lib:"
        )
  )))
```

I think the next step would be to query the nix store to automatically set those paths instead of hardcoding them like I did.

**Comment by spacekitteh** [comment]
Is there any update on this?

**Comment by mjlbach** [comment]
@spacekitteh Did the solution I post not work for you?

**Comment by spacekitteh** [comment]
Nope:
`Invalid function: (setenv "LIBRARY_PATH" (concat (shell-command-to-string "nix eval --raw '(let pkgs = import <nixpkgs> {}; in (pkgs.lib.getLib pkgs.libgccjit + /lib/gcc/x86_64-unknown-linux-gnu/9.3.0:)'") (shell-command-to-string "nix eval --raw '(let pkgs = import <nixpkgs> {}; in (pkgs.lib.getLib pkgs.stdenv.cc.cc + /lib)'") (shell-command-to-string "nix eval --raw '(let pkgs = import <nixpkgs> {}; in (pkgs.lib.getLib pkgs.stdenv.glibc + /lib)'")))`

**Comment by mjlbach** [comment]
Sorry, try this instead
```
(setq comp-async-env-modifier-form '((setenv "LIBRARY_PATH"
                                             (concat
        (shell-command-to-string "nix eval --raw '(let pkgs = import <nixpkgs> {}; in with pkgs; stdenv.lib.makeLibraryPath [stdenv.cc.cc stdenv.glibc])'")
        ":"
        (shell-command-to-string "nix eval --raw '(let pkgs = import <nixpkgs> {}; in with pkgs; lib.getLib libgccjit + /lib/gcc/x86_64-unknown-linux-gnu/9.3.0 )'")))))
```

**Comment by spacekitteh** [comment]
Nope, still getting `Invalid function`.

**Comment by mjlbach** [comment]
This is my output from ielm:
```
ELISP> (setenv "LIBRARY_PATH"
                                             (concat
        (shell-command-to-string "nix eval --raw '(let pkgs = import <nixpkgs> {}; in with pkgs; stdenv.lib.makeLibraryPath [stdenv.cc.cc stdenv.glibc])'")
        ":"
        (shell-command-to-string "nix eval --raw '(let pkgs = import <nixpkgs> {}; in with pkgs; lib.getLib libgccjit + /lib/gcc/x86_64-unknown-linux-gnu/9.3.0 )'")))
"/nix/store/a6z7ighixg7gb6krf9k60ylgmahij63x-gcc-9.3.0-lib/lib:/nix/store/jx19wa4xlh9n4324xdl9rjnykd19mmq3-glibc-2.30/lib:/nix/store/wfvfxj1w9gj21nbai6rsyhwvkbqf59w0-libgccjit-9.3.0/lib/gcc/x86_64-unknown-linux-gnu/9.3.0"
```

**Comment by spacekitteh** [comment]
Am I missing a `require` or anything?

**Comment by spacekitteh** [comment]
I use `use-package`, does that not play nicely or something?

**Comment by spacekitteh** [comment]
Could it be because I unfortunately don't use NixOS, but nix+home-manager on Ubuntu?

**Comment by mjlbach** [comment]
I'm using nix + home-manager on Fedora

**Comment by spacekitteh** [comment]
It works in `*scratch*`, but not when setting it as `comp-async-env-modifier-form`.

**Comment by mjlbach** [comment]
@adisbladis should this be reopened? I don't know when it happened, but my native-compilation has been silently failing via straight.el/doom even when setting LIBRARY_PATH

**Comment by adisbladis** [comment]
Sure, I'll reopen.
But I will not be debugging this myself as I only ever use Nix for Emacs package management.

**Comment by mjlbach** [comment]
Yes, I'm happy to look into this :)

**Comment by mjlbach** [comment]
This can now be closed, fix is pending merge into nixpkgs

**Comment by mjlbach** [comment]
Ok, so there issues is a bit more complicated.

@AndreaCorallo It seems like if you launch emacs with the correct LIBRARY_PATH, native compilation works fine:

```bash
LIBRARY_PATH=/nix/store/rnx85wkx3fdrwmj5x1s4h31jqr50l8s1-gcc-9.3.0-lib/lib:/nix/store/7p1v1b6ys9fydg5kdqvr5mpr8svhwf4p-glibc-2.31/lib:/nix/store/0dknchqiha9mvzhy6m3xmyzl3g2zc7av-libgccjit-9.3.0/lib/gcc/x86_64-unknown-linux-gnu/9.3.0  emacs
```

If you set comp-async-env-modifier-form to:

```elisp
(setq comp-async-env-modifier-form `(setenv "LIBRARY_PATH" "/nix/store/rnx85wkx3fdrwmj5x1s4h31jqr50l8s1-gcc-9.3.0-lib/lib:/nix/store/7p1v1b6ys9fydg5kdqvr5mpr8svhwf4p-glibc-2.31/lib:/nix/store/0dknchqiha9mvzhy6m3xmyzl3g2zc7av-libgccjit-9.3.0/lib/gcc/x86_64-unknown-linux-gnu/9.3.0"))
```

Native-compilation fails. Under the hood, the async-native-comp function is just launching a bunch of workers evaluating comp-async-env-modifier form before calling native-comp, however, the following fails to compile under nix. And just outputs empty *.eln files (with debugging on the c intermediate is output fine).

```elisp
(setenv "LIBRARY_PATH" "/nix/store/rnx85wkx3fdrwmj5x1s4h31jqr50l8s1-gcc-9.3.0-lib/lib:/nix/store/7p1v1b6ys9fydg5kdqvr5mpr8svhwf4p-glibc-2.31/lib:/nix/store/0dknchqiha9mvzhy6m3xmyzl3g2zc7av-libgccjit-9.3.0/lib/gcc/x86_64-unknown-linux-gnu/9.3.0")
(native-compile "/home/michael/.emacs.d/.local/straight/build/dash/dash.el")
```

It seems like we can't set the LIBRARY_PATH from within emacs for native-comp, so people will likely have to wrap emacs unless this is fixed upstream.

**Comment by AndreaCorallo** [comment]
Do you get any more specific error on the console or in the '*Async-native-compile-log*' buffer?

**Comment by mjlbach** [comment]
Yes, it's not finding the libgcc libraries it seems:

When I invoke directly:
```
/nix/store/k6rnzlmcvm37d0nw4fkz6ivmykmrdn39-binutils-2.31.1/bin/ld: cannot find crti.o: No such file or directory
/nix/store/k6rnzlmcvm37d0nw4fkz6ivmykmrdn39-binutils-2.31.1/bin/ld: cannot find crtbeginS.o: No such file or directory
/nix/store/k6rnzlmcvm37d0nw4fkz6ivmykmrdn39-binutils-2.31.1/bin/ld: cannot find -lgcc
/nix/store/k6rnzlmcvm37d0nw4fkz6ivmykmrdn39-binutils-2.31.1/bin/ld: cannot find -lgcc
libgccjit.so: error: error invoking gcc driver
```

Async native compiler log:

```
Loading /nix/store/c27zvaqcdjviss061zcb1mny5bncpsyv-emacs-gcc-20200726.0/share/emacs/site-lisp/site-start.el (source)...
Loading /nix/store/c27zvaqcdjviss061zcb1mny5bncpsyv-emacs-gcc-20200726.0/share/emacs/site-lisp/site-start.el (source)...
Loading /nix/store/c27zvaqcdjviss061zcb1mny5bncpsyv-emacs-gcc-20200726.0/share/emacs/site-lisp/site-start.el (source)...
Compiling /home/michael/.emacs.d/.local/straight/build/yasnippet/yasnippet.el...
Compiling /home/michael/.emacs.d/.local/straight/build/xterm-color/xterm-color.el...
Compiling /home/michael/.emacs.d/.local/straight/build/yasnippet/yasnippet-autoloads.el...
empty byte compiler output: "/home/michael/.emacs.d/.local/straight/build/yasnippet/yasnippet-autoloads.el", "/home/michael/.emacs.d/.local/straight/build/yasnippet/yasnippet-autoloads.el"
Loading /nix/store/c27zvaqcdjviss061zcb1mny5bncpsyv-emacs-gcc-20200726.0/share/emacs/site-lisp/site-start.el (source)...
Compiling /home/michael/.emacs.d/.local/straight/build/xterm-color/xterm-color-autoloads.el...
empty byte compiler output: "/home/michael/.emacs.d/.local/straight/build/xterm-color/xterm-color-autoloads.el", "/home/michael/.emacs.d/.local/straight/build/xterm-color/xterm-color-autoloads.el"
```

**Comment by AndreaCorallo** [comment]
Mmmh I see, this looks strange.  
So the exact same LIBRARY_PATH value if set before starting emacs does the job?

**Comment by mjlbach** [comment]
Yes, exactly. We can also wrap emacs so that it effectively passes the library_path on each startup but as @adisbladis pointed out this will make all subprocesses launches by emacs impure as they will then pass this library_env to all child processes

**Comment by jkachmar** [comment]
On the off chance that y'all may not have seen this, but it seems that when/if this enhancement lands it sounds like we should be able to sidestep this problem entirely:

https://debbugs.gnu.org/cgi/bugreport.cgi?bug=42761

**Comment by mjlbach** [comment]
Yeah, I was following. As I noted above, even if we set LIBRARY_PATH via `comp-async-env-modifier-form` we still can't native-compile packages (under async workers, sync doesn't invoke this so would not work to begin with). Currently LIBRARY_PATH is set incorrectly in nixpkgs, I have a PR to fix this, but as mentioned it doesn't solve the problem because gcc doesn't pick up the LIBRARY_PATH set from within emacs.

**Comment by jkachmar** [comment]
I might be misreading, but that thread seems to point to additionally exposed functionality (cf. [this diff](https://github.com/emacs-mirror/emacs/compare/feature/native-comp...antifuchs:allow-setting-driver-options)) which appears to allow for options to be passed _directly_ to the native compiler backend from within Emacs.

My understanding is that it would bypass the need to mess with `LIBRARY_PATH` entirely in this situation (although, again, I could be mistaken).

**Comment by matthew-piziak** [comment]
When I load `(setq comp-deferred-compilation t)` I simply get the following:
```
Warning (comp): No write access for /nix/store/kl5nkajk79p2wg8xiy4f354xmlxvhxis-emacs-packages-deps/share/emacs/site-lisp/elpa/use-package-20200721.2156/eln-x86_64-pc-linux-gnu-815168410c73dd5c/use-package-core.eln skipping. [91 times]
Warning (comp): No write access for /nix/store/kl5nkajk79p2wg8xiy4f354xmlxvhxis-emacs-packages-deps/share/emacs/site-lisp/elpa/bind-key-20200805.1727/eln-x86_64-pc-linux-gnu-815168410c73dd5c/bind-key.eln skipping. [5 times]
Warning (comp): No write access for /nix/store/kl5nkajk79p2wg8xiy4f354xmlxvhxis-emacs-packages-deps/share/emacs/site-lisp/elpa/use-package-20200721.2156/eln-x86_64-pc-linux-gnu-815168410c73dd5c/use-package-bind-key.eln skipping. [6 times]
Warning (comp): No write access for /nix/store/kl5nkajk79p2wg8xiy4f354xmlxvhxis-emacs-packages-deps/share/emacs/site-lisp/elpa/use-package-20200721.2156/eln-x86_64-pc-linux-gnu-815168410c73dd5c/use-package-diminish.eln skipping. [3 times]
Warning (comp): No write access for /nix/store/kl5nkajk79p2wg8xiy4f354xmlxvhxis-emacs-packages-deps/share/emacs/site-lisp/elpa/use-package-20200721.2156/eln-x86_64-pc-linux-gnu-815168410c73dd5c/use-package-delight.eln skipping. [3 times]
Warning (comp): No write access for /nix/store/kl5nkajk79p2wg8xiy4f354xmlxvhxis-emacs-packages-deps/share/emacs/site-lisp/elpa/use-package-20200721.2156/eln-x86_64-pc-linux-gnu-815168410c73dd5c/use-package-ensure.eln skipping. [7 times]
Warning (comp): No write access for /nix/store/kl5nkajk79p2wg8xiy4f354xmlxvhxis-emacs-packages-deps/share/emacs/site-lisp/elpa/delight-1.7/eln-x86_64-pc-linux-gnu-815168410c73dd5c/delight.eln skipping. [5 times]
Warning (comp): No write access for /nix/store/kl5nkajk79p2wg8xiy4f354xmlxvhxis-emacs-packages-deps/share/emacs/site-lisp/elpa/s-20180406.808/eln-x86_64-pc-linux-gnu-815168410c73dd5c/s.eln skipping. [69 times]
Warning (comp): No write access for /nix/store/kl5nkajk79p2wg8xiy4f354xmlxvhxis-emacs-packages-deps/share/emacs/site-lisp/elpa/exec-path-from-shell-20200526.324/eln-x86_64-pc-linux-gnu-815168410c73dd5c/exec-path-from-shell.eln skipping. [12 times]
Warning (comp): No write access for /nix/store/kl5nkajk79p2wg8xiy4f354xmlxvhxis-emacs-packages-deps/share/emacs/site-lisp/elpa/el-patch-20200716.1428/eln-x86_64-pc-linux-gnu-815168410c73dd5c/el-patch.eln skipping. [21 times]
Warning (comp): No write access for /nix/store/kl5nkajk79p2wg8xiy4f354xmlxvhxis-emacs-packages-deps/share/emacs/site-lisp/elpa/request-20200517.1305/eln-x86_64-pc-linux-gnu-815168410c73dd5c/request.eln skipping. [82 times]
Warning (comp): No write access for /nix/store/kl5nkajk79p2wg8xiy4f354xmlxvhxis-emacs-packages-deps/share/emacs/site-lisp/elpa/diminish-20191127.1326/eln-x86_64-pc-linux-gnu-815168410c73dd5c/diminish.eln skipping. [3 times]
…
```

**Comment by mjlbach** [comment]
This should be fixed by this change https://github.com/NixOS/nixpkgs/pull/94637/files#diff-66d54fc5596a4650473b8b4049520905R70-R75

closed by this PR https://github.com/NixOS/nixpkgs/pull/94637

**Comment by willbush** [comment]
For what it's worth, I'm getting the following error:

`Invalid function: (setenv "LIBRARY_PATH" (string-join (seq-filter (lambda (v) (null (eq v nil))) (list (getenv "LIBRARY_PATH") "/nix/store/y97bkww2pnxnqd2kwprlbw3ji4ib4d1w-libgccjit-9.3.0/lib/gcc/x86_64-unknown-linux-gnu/9.3.0:/nix/store/v589pqjhvxrj73g3r0xb41yr84z5pwb7-gcc-9.3.0-lib/lib:/nix/store/mh78fk3x12q2a77srgkzv16h0irl8r61-glibc-2.31/lib")) ":"))`

After I updated NixOS to the latest on the nixos-unstable channel and updated the emacs-overlay to the latest (currently rev `943ec43e54f5152c4d5d8d0076e72beea9a7ff5a` in the master branch).

**Comment by mjlbach** [comment]
The fix is still in staging; it will take a bit for it to end up in nixpkgs

**Comment by adisbladis** [comment]
This has now hit unstable!

Thank you @antifuchs!

**Comment by spacekitteh** [comment]
As of this morning, emacsGcc to no longer launches properly for me on NixOS. I'm not sure if it's related to this or not.

**Comment by willbush** [comment]
What error are you getting?

**Comment by spacekitteh** [comment]
Nevermind - I had to delete my ~/.emacs.d/eln-cache folder.
